### ISSUE TITLE
Add file-private `require`

### ISSUE THREAD

**Issue by asoffa** [crystal-lang/crystal#4368]
Hi,

Suppose I have a file that requires a large number of resources that are used for its internal behavior, e.g.
```
# in file sophisticated.cr:
require "a"
require "b"
...
require "z"

module Sophisticated
...
end
```

Now, any file that requires "sophisticated" will inherit the contents of all of the files a.cr, ..., z.cr that sophisticated.cr requires for its implementation **and** any requires within a.cr, ..., z.cr recursively. This can lead to code pollution, especially when there are modules, classes, etc. with similar names (and even redefinitions in the worst case if there are identical names). Therefore, it would be nice to have a new keyword (e.g. `use`) or some sort of marker for `require` (e.g. `private require`) so that the implementation details within sophisticated.cr are not leaked into files that merely *use* sophisticated.cr, e.g.
```
# in file sophisticated.cr:

# contents of a.cr, ..., z.cr now exposed to sophisticated.cr, but
# not to files that require/use it
use "a"
use "b"
...
use "z"

module Sophisticated
...
end
```
On the other hand, the leaking behavior is sometimes desirable for e.g. splitting up a module across multiple files, so it would be nice to have both the current behavior of `require` and this proposed addition.

This would be along the same lines as #3280, but for requiring files instead of defining modules, classes, constants, methods, etc.

Thoughts?

**Comment by RX14** [comment]
Duplicate of https://github.com/crystal-lang/crystal/issues/140. I don't think the core team's opinions have changed much.

**Comment by asoffa** [comment]
Hmm, it seems that the main issue mentioned in #140 was (from @asterite's comment):
"What @Nami-Doc says is the main reason Ruby doesn't implement this: monkey-patching. In any file you can reopen any other class. What would a file that reopen a class return?"

The `use` system I am proposing here would be identical to `require`, except that it wouldn't be recursive. For the case of a file reopening a class, it would behave as if the file were `require`d for any file that `use`s it and not `require`d for any file that doesn't `use` it (and no change from the current behavior if the file is instead `require`d).

The same monkey-patching "issue" (really a feature IMHO) also arises with `require` in a project such as the following:
```
# in file first.cr:
module A
  def self.first
    "first"
  end
end

# in file second.cr:
module A
  def self.second
    "second"
  end
end

# in file test.cr:
require "first"
A.first   #=> "first"
A.second  # error: doesn't work w/o `require "second"`
```
So the full A module isn't available in all cases anyway.

Now suppose a file `require`d by another file `require`d by yet another file that first.cr `require`s is updated so that it happens to `require` second.cr. Now the code in test.cr "mysteriously" works, and there's no way I can prevent it from working, even if I don't want `A.second` to be visible without the user explicitly asking for it with `require`/`use`

**Comment by asoffa** [comment]
Another issue: the order of `require`s. Suppose there's a library with
```
# in file first.cr
module A
  def self.first
    "first"
  end
end

# in file another_first.cr
module A
  def self.first
    "another_first"
  end
end
```
and the library user uses two files dependency1.cr and dependency2.cr that `require` files that `require` files that `require` first.cr and another_first.cr, respectively:
```
# in test.cr
require "dependency1"
require "dependency2"

Dependency1.do_your_thing_that_uses_first_very_indirectly  # "another_first" used
```
Now the user merely switch the order of the `require` statements:
```
# in test.cr
require "dependency2"
require "dependency1"

Dependency1.do_your_thing_that_uses_first_very_indirectly  # "first" used
```
This then leaves the user scratching his/her head, wondering why merely switching the order of the `require` statements caused completely different behavior that had seemingly nothing to do with `A.first` at all. (Perhaps the user didn't even know that module `A` existed behind the scenes.)

The above examples are perhaps a bit contrived, but hopefully illustrate why it would be useful to have `require` not be the **only** option for using code from another file

**Comment by ozra** [comment]
tl;dr - but I think this is about a feature I'd die for to have in Crystal, I'll just add my wording, and then you shoot me if it was straying from the subject.

- I'd want `require` for requiring files (duh! we've got it).
- One can include a module in the current module / "namespace", that's good when wanted, but:
- What I'd really really would like to have is `using` similar to in C++. Adding a module to "naked name lookups" within whatever narrow scope I choose (for instance maybe just within one branch of an if, that module should leave lookup-list when that scope ends), needless to say then: _without_ including it and polluting the namespace of my module (If the user of my module wants the features of a module mine is using - they'll include it them selves. I don't want to expose 3rd party modules just to be able to use them without namespace qualification.
If this is already possible, and I totally missed something here, I've been out of the loop to long, and I'm then also _very happy_. Please say it's so! :)

**Comment by asoffa** [comment]
@ozra What you propose above complements `extend`/`include`, whereas what I propose above complements `require`. But I agree that both would be useful

**Comment by asoffa** [comment]
@ozra By the way, what you propose above is (sort of) available in the following form:
```
module SomeReallyLongModuleName
  def self.sayHi
    puts "hi"
  end
end

module UsingSomeReallyLongModuleName
  # almost like the `using SomeReallyLongName` that you propose:
  private alias S = SomeReallyLongName

  def self.callSayHi
    S.sayHi  # still need to prefix with `S`, but this isn't nearly as cumbersome as
             # prefixing with `SomeReallyLongName`, especially if you have a lot of
             # calls to methods in `SomeReallyLongModuleName`
             # ...and it keeps clear where the `sayHi` method is coming from
  end
end
```
Now, if only something analogous were available with requiring files so that the *global* namespace didn't get polluted... :-)

**Comment by ozra** [comment]
@asoffa - thanks for the sharing of tricks - I feel it's "all or nothing" though :)

**Comment by asoffa** [comment]
#4439 (adding import hooks that allow a file to return a value such as a module or class instance without leaking all of the contents of the file) provides an alternative solution to the problem of encapsulating implementation details. If #4439 is accepted, this can be closed.

@ozra You may want to open a new ticket for your `using` proposal if it's something you would still like to see

**Comment by asoffa** [comment]
Adding from https://github.com/crystal-lang/crystal/issues/4439#issuecomment-303272821:

More evidence from other languages that `import` and `require` can play nicely together:
1. Python has both `import` and `execfile` (Python 2) / `exec` (Python 3)
2. Julia has both `import`/`using` and `include`
3. Nim has both `import` and `include`

Granted that in these languages the `import` way rather than the `require` way is the primary means of using code from other files, I have yet to see anyone complain about having both available

**Comment by ysbaddaden** [comment]
Closing as duplicate of #140. You may voice yourself in closed issues if you really have to, but plase don't duplicate issues.
